src/krb5/_principal.pyx

```diff
@@ -134,6 +283,63 @@ cdef class Principal:
         name = self.name
         return name.decode('utf-8') if name else 'NULL'
 
+    @property
+    def realm(Principal self) -> bytes:
+        cdef size_t length
+        cdef char *value
+
+        if not self.raw:
+            raise ValueError("Attempting to access property of NULL principal")
+
+        pykrb5_principal_get(self.raw, &length, &value, NULL, NULL)
+
+        if length == 0:
+            return b""
+        else:
+            return value[:length]
+
+    @property
+    def components(Principal self) -> typing.List[bytes]:
+        cdef size_t component_count
+
+        cdef size_t length
+        cdef char *value
+
+        if not self.raw:
+            raise ValueError("Attempting to access property of NULL principal")
+
+        pykrb5_principal_get(self.raw, NULL, NULL, &component_count, NULL)
+
+        components = []
+        for pos in range(component_count):
+            pykrb5_principal_get_component(self.raw, pos, &length, &value)
+
+            if length == 0:
+                component = b""
+            else:
+                component = value[:length]
+
+            components.append(component)
+
+        return components
+
+    @property
+    def type(Principal self) -> int:
```

Comment:
Are these well known constants between both MIT and Heimdal. I'm wondering whether we should be returning an enum here.

Reply:
These are the name types from RFC 4120: <https://www.rfc-editor.org/rfc/rfc4120#section-7.5.8>

The problem with an enum would be that if the library gets a principal name from somewhere else which has a name types which is not in the list, it will throw an exception when attempting to read the `type` accessor, because python enums don't allow unknown values.

Reply:
You can definitely still use an enum, it just needs to define a custom [_missing_](https://docs.python.org/3/library/enum.html#enum.Enum._missing_) implementation to define it.

```python
import enum

class MyEnum(enum.IntEnum):
    VALUE1 = 1
    VALUE2 = 2

    @classmethod
    def _missing_(cls, value: object) -> typing.Optional[enum.Enum]:
        if not isinstance(value, int):
            return None

        new_member = int.__new__(cls, value)
        new_member._name_ = f"Unknown MyEnum 0x{value:08X}"
        new_member._value_ = value
        return cls._value2member_map_.setdefault(value, new_member)


value = MyEnum(3)
str(value)  # '3'  - Depends on python version though
int(value)  # 3
value.name  # 'Unknown MyEnum 0x00000003'
value.value  # 3
```

Granted `_value2member_map_` may not be documented but I think in this case we can be mostly sure that the enum values that are returned should stay the same and this fallback is more of a just in case situation.

Reply:
I've added such an enum (`krb5.NameType`).